### PR TITLE
Show progress for parent processes

### LINKED ISSUE #4805: Show progress summary in parent process
Currently, for a superordinated process, no progress bar is displayed in the process table. The space for it remains empty.

**Goal:** A progress bar, which shows the cumulative process progress of the subordinate processes, is to be displayed here.

_This development is funded by the Municipality of The Hague—Construction Office._

### PR DESCRIPTION
**Pull request by matthias-ronge**
Fixes #4805

### CODE REVIEW

File: Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java [hunk 1]
```diff
@@ -920,6 +921,27 @@ private void convertRelatedJSONObjects(Map<String, Object> jsonObject, ProcessDT
...
+                processDTO.setProgressInProcessing(progressOfChildren(children, tasks -> getProgressInProcessing(tasks, null)));
+                processDTO.setProgressOpen(progressOfChildren(children, tasks -> getProgressOpen(tasks, null)));
+                processDTO.setProgressLocked(progressOfChildren(children, tasks -> getProgressLocked(tasks, null)));
+            } catch (DAOException dao) {
+                throw new DataException(dao);
+            }
+        }
+    }
+
+    private static Double progressOfChildren(List<Process> children, Function<List<Task>, Double> calculator) {
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java:939]: Why are you handing the function here?
Please use:
`private Double progressOfChildren(List<Process> children) {
        DescriptiveStatistics statistics = new DescriptiveStatistics();
        for (Process child : children) {
            statistics.addValue(getProgressLocked(child.getTasks(), null));
        }
        return statistics.getMean();
    }`

File: Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java [hunk 2]
```diff
@@ -18,6 +18,14 @@
 import static org.kitodo.data.database.enums.CorrectionComments.NO_OPEN_CORRECTION_COMMENTS;
 import static org.kitodo.data.database.enums.CorrectionComments.OPEN_CORRECTION_COMMENTS;
 
+import com.itextpdf.text.Document;
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java:21]: There seems to be some mixup with the imports here

File: Kitodo/src/main/webapp/WEB-INF/templates/includes/processes/progressColumn.xhtml [hunk 3]
```diff
@@ -19,7 +19,7 @@
     <h:panelGroup
             class="process-progress"
             title="#{ProcessForm.getCurrentTaskTitles(process)}"
-            rendered="#{not empty process.tasks}"
+            rendered="#{not empty process.tasks or process.hasChildren()}"
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/webapp/WEB-INF/templates/includes/processes/progressColumn.xhtml:22]: What happens, if the process has children, but they don't have tasks? (e.g year processes)

**Review by Kathrin-Huber — CHANGES_REQUESTED** [review]

File: Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java (hunk 1 above)
Comment (matthias-ronge) [reply Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java:938]: I wanted to deduplicate the code. What you want gives 4× the same function, just differing in the function called:

```java
            processDTO.setProgressClosed(progressOfChildrenClosed(children));
            processDTO.setProgressInProcessing(progressOfChildrenInProcessing(children));
            processDTO.setProgressOpen(progressOfChildrenOpen(children));
            processDTO.setProgressLocked(progressOfChildrenLocked(children));
        } catch (DAOException dao) {
            throw new DataException(dao);
        }
    }
}

private Double progressOfChildrenClosed(List<Process> children) {
    DescriptiveStatistics statistics = new DescriptiveStatistics();
    for (Process child : children) {
        statistics.addValue(getProgressClosed(child.getTasks(), null));
    }
    return statistics.getMean();
}

private Double progressOfChildrenInProcessing(List<Process> children) {
    DescriptiveStatistics statistics = new DescriptiveStatistics();
    for (Process child : children) {
        statistics.addValue(getProgressInProcessing(child.getTasks(), null));
    }
    return statistics.getMean();
}

private Double progressOfChildrenOpen(List<Process> children) {
    DescriptiveStatistics statistics = new DescriptiveStatistics();
    for (Process child : children) {
        statistics.addValue(getProgressOpen(child.getTasks(), null));
    }
    return statistics.getMean();
}

private Double progressOfChildrenLocked(List<Process> children) {
    DescriptiveStatistics statistics = new DescriptiveStatistics();
    for (Process child : children) {
        statistics.addValue(getProgressLocked(child.getTasks(), null));
    }
    return statistics.getMean();
}
```

Do you really want that?

File: Kitodo/src/main/webapp/WEB-INF/templates/includes/processes/progressColumn.xhtml (hunk 3 above)
Comment (solth) [reply Kitodo/src/main/webapp/WEB-INF/templates/includes/processes/progressColumn.xhtml]: This does not seem to work as intended, see #5171

**Pull request merged**